File: animate_optitrack.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
import math

# === Konfiguracja / Wczytanie ===
filename = 'OptitrackAgataPassiveWAggressive3.csv'
frameRate = 100

raw = pd.read_csv(filename, header=None).values.astype(float)
raw[np.isinf(raw)] = np.nan
raw[raw == 0] = np.nan

size_skeleton = raw.shape
skeleton = np.zeros((size_skeleton[0]-6, 65), dtype=float)

skeleton[:,0:2] = raw[6:, 0:2]
skeleton[:,2:5]   = raw[6:, 237:240]  #1
skeleton[:,5:8]   = raw[6:, 224:227]  #2
skeleton[:,8:11]  = raw[6:, 211:214]  #3
skeleton[:,11:14] = raw[6:, 172:175]  #4
skeleton[:,14:17] = raw[6:, 185:188]  #5
skeleton[:,17:20] = raw[6:, 198:201]  #6
skeleton[:,20:23] = (raw[6:,211:214] + raw[6:,172:175]) / 2.0  #7
skeleton[:,23:26] = (((raw[6:,126:129] + raw[6:,129:132]) / 2.0) + ((raw[6:,77:80] + raw[6:,80:83]) / 2.0)) / 2.0  #8
skeleton[:,26:29] = raw[6:,58:61]   #9
skeleton[:,29:32] = raw[6:,61:64]   #10
skeleton[:,32:35] = raw[6:,156:159] #11
skeleton[:,35:38] = (raw[6:,139:142] + raw[6:,142:145]) / 2.0  #12
skeleton[:,38:41] = (raw[6:,126:129] + raw[6:,129:132]) / 2.0  #13
skeleton[:,41:44] = (raw[6:,77:80] + raw[6:,80:83]) / 2.0      #14
skeleton[:,44:47] = (raw[6:,90:93] + raw[6:,93:96]) / 2.0      #15
skeleton[:,47:50] = raw[6:,107:110]  #16
skeleton[:,50:53] = raw[6:,371:374]  #17
skeleton[:,53:56] = raw[6:,374:377]  #18
skeleton[:,56:59] = raw[6:,377:380]  #19
skeleton[:,59:62] = raw[6:,380:383]  #kamera (nr 20 w MATLABie)

# Dodaj punkt 21 (fixed) jako [0,0,0] (w oryginale wpisują do 51:53)
fixedPoint21 = np.tile(np.array([0.0,0.0,0.0]), (skeleton.shape[0],1))
skeleton[:,50:53] = fixedPoint21

nFrames = skeleton.shape[0]
dataOnly = skeleton[:, 2:]  # od trzeciej kolumny
nMarkers = dataOnly.shape[1] // 3

# przygotuj strukturę Data.markers [nMarkers x 3 x nFrames]
Data_markers = np.full((nMarkers, 3, nFrames), np.nan)
for f in range(nFrames):
    frameData = dataOnly[f,:]
    for m in range(nMarkers):
        idx = 3*m
        Data_markers[m, :, f] = frameData[idx:idx+3]

# Workspace vertices (kopiowane z MATLABa)
workspaceVertices = np.array([
    [0, 0, -0.5],
    [0, 0, 3.6],
    [4, 0, 3.6],
    [4, 2.5, 3.6],
    [0, 2.5, 3.6],
    [4, 0, -0.5],
    [0, 2.5, -0.5],
    [4, 2.5, -0.5],
    [0, 0, 1.53],
    [4, 0, 1.53],
    [0, 2.5, 1.53],
    [4, 2.5, 1.53],
])

workspaceEdges = np.array([
    [101,102],[102,103],[103,104],[104,105],[105,102],
    [107,108],[108,106],
    [101,106],[101,107],[103,106],[104,108],[105,107],
    [109,110],[110,112],[111,112],[111,109]
])

# Rysowanie (Matplotlib 3D)
fig = plt.figure(figsize=(9,6))
ax = fig.add_subplot(111, projection='3d')
plt.ion()
ax.set_box_aspect([1,1,1])

# Pierwsza klatka
M0 = Data_markers[:,:,0]
scatter = ax.scatter(M0[:,0], M0[:,1], M0[:,2], s=50)

# Linie szkieletu
skeletonLines = np.array([
    [1,2],[2,3],[3,7],[4,7],[4,5],[5,6],
    [7,8],[8,9],[9,10],
    [8,13],[13,12],[12,11],
    [8,14],[14,15],[15,16]
], dtype=int)

line_handles = []
for pair in skeletonLines:
    i1 = pair[0]-1
    i2 = pair[1]-1
    ln, = ax.plot([M0[i1,0], M0[i2,0]],
                  [M0[i1,1], M0[i2,1]],
                  [M0[i1,2], M0[i2,2]],
                  'g-', linewidth=2)
    line_handles.append(ln)

# Etykiety markerów
label_texts = []
for m in range(nMarkers):
    t = ax.text(M0[m,0], M0[m,1], M0[m,2], str(m+1), fontsize=8, verticalalignment='bottom')
    label_texts.append(t)

# Wierzchołki workspace
ax.scatter(workspaceVertices[:,0], workspaceVertices[:,1], workspaceVertices[:,2], s=60, c=(0,0.4,0))

# Etykiety workspace 101-112
for i, v in enumerate(workspaceVertices):
    ax.text(v[0], v[1], v[2], str(100+i+1), color=(0,0.4,0), fontsize=8)

# Linie workspace
for edge in workspaceEdges:
    idx1 = edge[0]-101
    idx2 = edge[1]-101
    v1 = workspaceVertices[idx1]
    v2 = workspaceVertices[idx2]
    ax.plot([v1[0], v2[0]], [v1[1], v2[1]], [v1[2], v2[2]], color=(0,0.4,0), linewidth=2)

# Promienie z punktu kamery (nr 20 w MATLAB => cameraIdx=20 -> tu marker index 19)
cameraIdx = 19
horizontalFOV = math.radians(82)
verticalFOV = math.radians(52)
rayLength = 5.5
leftRayDir = np.array([-math.sin(horizontalFOV/2), 0, -math.cos(horizontalFOV/2)])
rightRayDir = np.array([ math.sin(horizontalFOV/2), 0, -math.cos(horizontalFOV/2)])
topRayDir = np.array([0, math.sin(verticalFOV/2), -math.cos(verticalFOV/2)])
bottomRayDir = np.array([0, -math.sin(verticalFOV/2), -math.cos(verticalFOV/2)])
violet = (0.5, 0, 0.5)

cameraPoint = M0[cameraIdx,:]
ray_lines = []
for d in [leftRayDir, rightRayDir, topRayDir, bottomRayDir]:
    ln, = ax.plot([cameraPoint[0], cameraPoint[0]+rayLength*d[0]],
                  [cameraPoint[1], cameraPoint[1]+rayLength*d[1]],
                  [cameraPoint[2], cameraPoint[2]+rayLength*d[2]],
                  color=violet, linewidth=2)
    ray_lines.append(ln)

ax.set_xlabel('X'); ax.set_ylabel('Y'); ax.set_zlabel('Z')
ax.set_title('Animacja optitrack')

# Ustawienia osi (dopasuj do danych)
ax.auto_scale_xyz([-1,5], [-1,3], [-1,4])

# Animacja
for f in range(nFrames):
    M = Data_markers[:,:,f]
    # update scatter
    scatter._offsets3d = (M[:,0], M[:,1], M[:,2])
    # update labels
    for m in range(nMarkers):
        label_texts[m].set_position((M[m,0], M[m,1]))
        label_texts[m].set_3d_properties(M[m,2], zdir='z')
    # update skeleton lines
    for idx, pair in enumerate(skeletonLines):
        i1 = pair[0]-1
        i2 = pair[1]-1
        x = [M[i1,0], M[i2,0]]
        y = [M[i1,1], M[i2,1]]
        z = [M[i1,2], M[i2,2]]
        line_handles[idx].set_data(x, y)
        line_handles[idx].set_3d_properties(z, zdir='z')
    # update rays (kamera)
    cameraPoint = M[cameraIdx,:]
    dirs = [leftRayDir, rightRayDir, topRayDir, bottomRayDir]
    for ri in range(4):
        d = dirs[ri]
        xs = [cameraPoint[0], cameraPoint[0] + rayLength*d[0]]
        ys = [cameraPoint[1], cameraPoint[1] + rayLength*d[1]]
        zs = [cameraPoint[2], cameraPoint[2] + rayLength*d[2]]
        ray_lines[ri].set_data(xs, ys)
        ray_lines[ri].set_3d_properties(zs, zdir='z')

    ax.set_title(f'Klatka {f+1} / {nFrames}')
    plt.draw()
    plt.pause(0.01)  # kontrola prędkości animacji

plt.ioff()
plt.show()

File: animate_optitrack.m
close all; clc;

filename = 'Take 2025-07-28 02.37.20 PM.csv';
frameRate = 100;

raw = readmatrix(filename);
raw(isinf(raw)) = NaN;
raw(raw == 0) = NaN;
size_skeleton = size(raw);
skeleton = zeros(size_skeleton(1) - 6, 65);

%%% szkielet agata
skeleton(:,1:2) = raw(7:end,1:2);
skeleton(:,3:5) = raw(7:end,238:240); %1
skeleton(:,6:8) = raw(7:end,225:227); %2
skeleton(:,9:11) = raw(7:end,212:214);%3
skeleton(:,12:14) = raw(7:end,173:175);%4
skeleton(:,15:17) = raw(7:end,186:188);%5
skeleton(:,18:20) = raw(7:end,199:201);%6
skeleton(:,21:23) = (raw(7:end,212:214) + raw(7:end,173:175))/2;%7
skeleton(:,24:26) = (((raw(7:end,127:129) + raw(7:end,130:132))/2) + ((raw(7:end,78:80) + raw(7:end,81:83))/2))/2;%8
skeleton(:,27:29) = raw(7:end,59:61);%9
skeleton(:,30:32) = raw(7:end,62:64);%10
skeleton(:,33:35) = raw(7:end,157:159);%11
skeleton(:,36:38) = (raw(7:end,140:142) + raw(7:end,143:145))/2;%12 
skeleton(:,39:41) = (raw(7:end,127:129) + raw(7:end,130:132))/2;%13
skeleton(:,42:44) = (raw(7:end,78:80) + raw(7:end,81:83))/2;%14
skeleton(:,45:47) = (raw(7:end,91:93) + raw(7:end,94:96))/2;%15
skeleton(:,48:50) = raw(7:end,108:110);%16
skeleton(:,51:53) = raw(7:end,372:374); % punkt stały 17
skeleton(:,54:56) = raw(7:end,375:377); % punkt stały 18
skeleton(:,57:59) = raw(7:end,378:380); % punkt stały 19
skeleton(:,60:62) = raw(7:end,366:368); % kamera

% Dodajemy punkt 21 (nr 21) na stałe: [0,0,0]
fixedPoint21 = repmat([0,0,0], size(skeleton,1), 1);
skeleton(:,51:53) = fixedPoint21;

% Liczba klatek i markerów
nFrames = size(skeleton, 1);
dataOnly = skeleton(:, 3:end);
nMarkers = floor(size(dataOnly, 2) / 3);

% Struktura danych
Data.nFrames = nFrames;
Data.nMarkers = nMarkers;
Data.markers = NaN(nMarkers, 3, nFrames);

for f = 1:nFrames
    frameData = dataOnly(f, :);
    for m = 1:nMarkers
        idx = (m - 1) * 3 + 1;
        Data.markers(m, :, f) = frameData(idx:idx+2);
    end
end

% Wierzchołki workspace: 101–108
workspaceVertices = [
    0, 0, -0.5;     % 101
    0, 0, 3.6;       % 102
    4, 0, 3.6;       % 103
    4, 2.5, 3.6;      % 104
    0, 2.5, 3.6;     % 105
    4, 0, -0.5;      % 106
    0, 2.5, -0.5;    % 107
    4, 2.5, -0.5    % 108
    0, 0, 1.53      %109
    4, 0, 1.53      %110
    0, 2.5, 1.53    %111
    4, 2.5, 1.53    %112
];

% Krawędzie workspace
workspaceEdges = [
    101 102; 102 103; 103 104; 104 105; 105 102;         
    107 108; 108 106;  
    101 106; 101 107; 103 106; 104 108; 105 107  ;
    109 110; 110 112; 111 112; 111 109;
];

figure; hold on;

% Pierwsza klatka szkieletu
M0 = squeeze(Data.markers(:,:,1));
scatterObj = scatter3(M0(:,1), M0(:,2), M0(:,3), 50, 'filled');

% Linie szkieletu
skeletonLines = [
    1 2; 2 3; 3 7; 4 7; 4 5; 5 6;
    7 8; 8 9; 9 10;
    8 13; 13 12; 12 11;
    8 14; 14 15; 15 16
];

lineHandles = gobjects(size(skeletonLines,1),1);
for i = 1:size(skeletonLines,1)
    idx1 = skeletonLines(i,1);
    idx2 = skeletonLines(i,2);
    lineHandles(i) = plot3([M0(idx1,1) M0(idx2,1)], ...
                           [M0(idx1,2) M0(idx2,2)], ...
                           [M0(idx1,3) M0(idx2,3)], ...
                           'g-', 'LineWidth', 2);
end

% Etykiety markerów
labelHandles = gobjects(nMarkers,1);
for m = 1:nMarkers
    labelHandles(m) = text(M0(m,1), M0(m,2), M0(m,3), ...
        num2str(m), 'VerticalAlignment', 'bottom', 'FontSize', 8);
end

% Wierzchołki workspace
scatter3(workspaceVertices(:,1), workspaceVertices(:,2), workspaceVertices(:,3), ...
    60, [0 0.4 0], 'filled');

% Etykiety workspace: 101–108
for i = 1:size(workspaceVertices,1)
    text(workspaceVertices(i,1), workspaceVertices(i,2), workspaceVertices(i,3), ...
        num2str(100+i), 'Color', [0 0.4 0], 'FontSize', 8);
end

% Linie workspace
for i = 1:size(workspaceEdges,1)
    idx1 = workspaceEdges(i,1)-100; 
    idx2 = workspaceEdges(i,2)-100;
    plot3([workspaceVertices(idx1,1), workspaceVertices(idx2,1)], ...
          [workspaceVertices(idx1,2), workspaceVertices(idx2,2)], ...
          [workspaceVertices(idx1,3), workspaceVertices(idx2,3)], ...
          'Color',[0 0.4 0],'LineWidth',2);
end

% --- DODANE: promienie z punktu kamery (nr 20)
cameraIdx = 20;  
horizontalFOV = deg2rad(82);
verticalFOV = deg2rad(52);
rayLength = 5.5;
leftRayDir = [ -sin(horizontalFOV/2), 0, -cos(horizontalFOV/2) ];
rightRayDir = [ sin(horizontalFOV/2), 0, -cos(horizontalFOV/2) ];
topRayDir = [ 0, sin(verticalFOV/2), -cos(verticalFOV/2) ];
bottomRayDir = [ 0, -sin(verticalFOV/2), -cos(verticalFOV/2) ];

violet = [0.5 0 0.5];

cameraPoint = M0(cameraIdx,:);
rayHandles(1) = plot3([cameraPoint(1), cameraPoint(1)+rayLength*leftRayDir(1)], ...
    [cameraPoint(2), cameraPoint(2)+rayLength*leftRayDir(2)], ...
    [cameraPoint(3), cameraPoint(3)+rayLength*leftRayDir(3)], 'Color', violet, 'LineWidth', 2);
rayHandles(2) = plot3([cameraPoint(1), cameraPoint(1)+rayLength*rightRayDir(1)], ...
    [cameraPoint(2), cameraPoint(2)+rayLength*rightRayDir(2)], ...
    [cameraPoint(3), cameraPoint(3)+rayLength*rightRayDir(3)], 'Color', violet, 'LineWidth', 2);
rayHandles(3) = plot3([cameraPoint(1), cameraPoint(1)+rayLength*topRayDir(1)], ...
    [cameraPoint(2), cameraPoint(2)+rayLength*topRayDir(2)], ...
    [cameraPoint(3), cameraPoint(3)+rayLength*topRayDir(3)], 'Color', violet, 'LineWidth', 2);
rayHandles(4) = plot3([cameraPoint(1), cameraPoint(1)+rayLength*bottomRayDir(1)], ...
    [cameraPoint(2), cameraPoint(2)+rayLength*bottomRayDir(2)], ...
    [cameraPoint(3), cameraPoint(3)+rayLength*bottomRayDir(3)], 'Color', violet, 'LineWidth', 2);

grid on;
xlabel('X'); ylabel('Y'); zlabel('Z');
axis equal;

% Animacja
for f = 1:Data.nFrames
    M = squeeze(Data.markers(:,:,f));
    set(scatterObj, 'XData', M(:,1), 'YData', M(:,2), 'ZData', M(:,3));
    for m = 1:nMarkers
        set(labelHandles(m), 'Position', M(m,:));
    end
    for i = 1:size(skeletonLines,1)
        idx1 = skeletonLines(i,1);
        idx2 = skeletonLines(i,2);
        set(lineHandles(i), ...
            'XData', [M(idx1,1), M(idx2,1)], ...
            'YData', [M(idx1,2), M(idx2,2)], ...
            'ZData', [M(idx1,3), M(idx2,3)]);
    end
    
    % Aktualizuj promienie
    cameraPoint = M(cameraIdx,:);
    set(rayHandles(1), 'XData', [cameraPoint(1), cameraPoint(1)+rayLength*leftRayDir(1)], ...
        'YData', [cameraPoint(2), cameraPoint(2)+rayLength*leftRayDir(2)], ...
        'ZData', [cameraPoint(3), cameraPoint(3)+rayLength*leftRayDir(3)]);
    set(rayHandles(2), 'XData', [cameraPoint(1), cameraPoint(1)+rayLength*rightRayDir(1)], ...
        'YData', [cameraPoint(2), cameraPoint(2)+rayLength*rightRayDir(2)], ...
        'ZData', [cameraPoint(3), cameraPoint(3)+rayLength*rightRayDir(3)]);
    set(rayHandles(3), 'XData', [cameraPoint(1), cameraPoint(1)+rayLength*topRayDir(1)], ...
        'YData', [cameraPoint(2), cameraPoint(2)+rayLength*topRayDir(2)], ...
        'ZData', [cameraPoint(3), cameraPoint(3)+rayLength*topRayDir(3)]);
    set(rayHandles(4), 'XData', [cameraPoint(1), cameraPoint(1)+rayLength*bottomRayDir(1)], ...
        'YData', [cameraPoint(2), cameraPoint(2)+rayLength*bottomRayDir(2)], ...
        'ZData', [cameraPoint(3), cameraPoint(3)+rayLength*bottomRayDir(3)]);
    
    title(sprintf('Klatka %d / %d', f, Data.nFrames));
    drawnow limitrate;
end
